array_methods_lab.js

```diff
@@ -0,0 +1,365 @@
+
+// ```
+// Input: [6,7,8,9]
+// Output: [0,7,16,27]
+// ```
+const numberTimesIdx = (arr) => {
+    return arr.map((el, i) => {
+        return el * i;
+    })
+}
+console.log(numberTimesIdx([6,7,8,9]));
+
+// 10. Write a function that returns whether or not every value in an array is a positive even number that doesn't end in 0.
+
+// ```
+// Input: [2,4,6,8,12]
+// Output: true
+
+// Input: [2,4,6,8,10]
+// Output: false
+// ```
+const posEvenNum = (arr) => {
+    return arr.every((el)=>{
+        if(el > 0 && el % 10 !== 0 ){
+            return el;
+        }
+    })
+}
+console.log(posEvenNum([2,4,6,8,12]));  // true
+console.log(posEvenNum([2,4,6,8,10]));  // false
+
+// 11. Write a function that returns a new array containing all of the strings in the original array uppercased.
+
+// ```
+// Input: [{}, 1, "cat", 3, ["hi"], {name: "dog"}, "dog", "bear"]
+// Output: ["CAT", "DOG", "BEAR"]
+// ```
+const upperCase = arr => {
+    let newArr = arr.filter((el)=>{
+        if(typeof el === 'string'){
+            return el;
+        }
+    })
+    return newArr.map((el)=>{
+        return el.toUpperCase();
+    })
+}
+console.log(upperCase([{}, 1, "cat", 3, ["hi"], {name: "dog"}, "dog", "bear"]));
+
+// ## Bonus:
+
+// You can write your own array methods!
+
+// ```js
+Array.prototype.sayHi = function() {
+    console.log(this)
+    console.log("Hi")
+}
+// ```
+
+// Now every array that you write has a `sayHi` method that you can call.
+
+// ```js
+let myArr = [1,2,3]
+myArr.sayHi()   //Logs "Hi" to the console
+// ```
+
+
+// 1. Using Array.prototype add a new method `myForEach` that mimics the behavior of the built in `forEach`.
+Array.prototype.myForEach = function(cb){
+    for(let i = 0; i < this.length; i++){
+        cb(this[i]);
+    }
+}
+
+
+// 2. Use your `myForEach` to add `myMap` to the Array prototype. `myMap` should behave the same as regular `map`.
+Array.prototype.myMap = function(cb){
+    let result = [];
+    for (let i = 0; i < this.length; i++){
+        result.push(this[i]);
+    }
+    return result;
+}
+
+
+
+// 3. Write a `myFilter` that behaves the same as `filter`.
+Array.prototype.myFilter = function(callback){
+    let result = [];
+    for(let i = 0; i < this.length; i++){
+        if(callback(this[i])){
+            result.push(this[i]);
+        } 
+    }
+    return result;
+}
+
+
+
+// 4. Write a `myEvery` that behaves the same as `every`.
+Array.prototype.myEvery = function (callback){
+    let bool = true;
+    for(let i = 0; i < this.length; i++){
+        if(!callback(this[i])){
+            bool = false;
+        }
+    }
+    return bool;
+}
+
+
+// 5. Write a `myReduce`. It should use the first element in the array if none is given.
+Array.prototype.myReduce = function (cb){
+    let acc = initialValue || undefined;
+    for(let i = 0; i < this.length; i++){
+        if (acc){
+            acc = callback.call(acc,acc,this[i], i, this);
```

Comment:
Looks like you copy pasta'd this =_=

acc = cb(acc,this[i], i, this) <---- This is what you wanted.


Keep working on this one, you might get asked this in a job interview, you never know.